cognite/powerops/resync/v1/config_to_fdm.py

```diff
@@ -0,0 +1,528 @@
+
+        Raises:
+            ValueError: If any required fields are missing or invalid for the specified data type.
+        """
+
+        logger.info(f"Processing {raw_data} as type {type_configuration.type_.__name__}")
+
+        # check if any fields from input data don't match the domain model
+        check_input_keys(raw_data, list(type_configuration.properties.keys()))
+
+        # split data into immutable, property configuration and reference data
+        immutable_data, property_configuration_data, reference_data = self._split_data_by_parsing_method(raw_data)
+
+        # populate properties from configurations
+        parsed_properties, generated_data_model_objects = self._populate_properties_from_configurations(
+            property_configuration_data, type_configuration.property_configurations, raw_data
+        )
+        data_model_objects.update(generated_data_model_objects)
+
+        # populate all external id references
+        parsed_reference_data, reference_external_ids = parse_external_ids(
+            reference_data, type_configuration.properties, self.resync_configuration.all_write_classes
+        )
+        external_ids.extend(reference_external_ids)
+
+        # join all dictionaries together in priority order
+        parsed_data = self._join_data_by_parsing_method(immutable_data, parsed_properties, parsed_reference_data)
+
+        # TODO: handle these special cases in a better way
+        if type_configuration.type_ is v1_data_classes.PlantWaterValueBasedWrite:
+            # TODO: remove when model is updated to a list instead of json
+            raw_penstock_head_loss_factor = parsed_data["penstock_head_loss_factors"]
+            parsed_data["penstock_head_loss_factors"] = {
+                str(index): float(loss_factor)
+                for index, loss_factor in enumerate(raw_penstock_head_loss_factor, start=1)
+            }
+
+            parsed_data["generators"] = (
+                self._get_all_connections(parsed_data, "plant", "generator")
+                if "generators" not in parsed_data
+                else parsed_data["generators"]
+            )
+            parsed_data["connection_losses"] = (
+                self._get_connection_losses(parsed_data)
+                if "connection_losses" not in parsed_data
+                else parsed_data["connection_losses"]
+            )
+
+        # TODO: extract all sub objects into the data_model_objects list
+        try:
+            fdm_object = type_configuration.type_(**parsed_data)
+            data_model_objects[fdm_object.external_id] = fdm_object
+        except ValidationError as exc:
+            raise ValueError(
+                f"Missing/invalid field for {type_configuration.type_.__name__} in data {parsed_data}"
+            ) from exc
+
+    @staticmethod
+    def _split_data_by_parsing_method(raw_data: dict) -> tuple[dict[str, Any], dict[str, Any], dict[str, Any]]:
```

Comment:
Any reason not to be as specific about the type of the raw_data argument as for the return types?
```suggestion
    def _split_data_by_parsing_method(raw_data: dict[str, Any]) -> tuple[dict[str, Any], dict[str, Any], dict[str, Any]]:
```